### PR TITLE
BUG: Updates to thwart failed message sends if batch size > 256Kib

### PR DESCRIPTION
**Pull request by malimohub**
@jwcorle great library man. We are using this heavily. Something popped up during load testing as outlined  below, let me know if you have any questions or want any updates:

Based on the AWS documentation (https://docs.aws.amazon.com/AWSSimpleQueueService/latest/APIReference/API_SendMessageBatch.html) the current batch message size constraints are as follows:
"The maximum allowed
// individual message size and the maximum total payload size (the sum of the individual lengths
// of all of the batched messages) are both 256 KiB (262,144 bytes)"

Currently s3 functionality is trggered if any message within the batch >  256 KiB, Which causes an error when sending a batch of messages where each are below the 256Kib default threshold (e.g a batch of 10 messages each with a size of 30Kib) but sum to > 256Kib

### CODE REVIEW

**Review by jwcorle — CHANGES_REQUESTED** [review]
Thanks @malimohub, glad you are using it and appreciate you contributing back this to the library! I made a couple small syntax suggestions for now, but after reading through I think there is still something missing with the update to `SendMessageBatch`.

There's a couple scenarios here that we'd like to cover:
1. If the entire **batch** size exceeds 256 KB, then send **ALL** records to S3 (in the future this might be tweaked to only send the messages that cause it to exceed this limit). You appear to cover this scenario already!
2. If an **individual** message exceeds the `messageSizeThreshold` (which can be configured), then that **individual** message should be sent to S3 as a pointer. This behavior I believe is slightly muddied because you compare this `messageSizeThreshold` to the `batchMsgSize` (`batchMsgSize < c.messageSizeThreshold`). This comparison I don't think will handle this scenario as nicely.

Could you make the tweaks to support this second scenario?

**Comment by malimohub** [comment]
> Thanks @malimohub, glad you are using it and appreciate you contributing back this to the library! I made a couple small syntax suggestions for now, but after reading through I think there is still something missing with the update to `SendMessageBatch`.
> 
> There's a couple scenarios here that we'd like to cover:
> 
> 1. If the entire **batch** size exceeds 256 KB, then send **ALL** records to S3 (in the future this might be tweaked to only send the messages that cause it to exceed this limit). You appear to cover this scenario already!
> 2. If an **individual** message exceeds the `messageSizeThreshold` (which can be configured), then that **individual** message should be sent to S3 as a pointer. This behavior I believe is slightly muddied because you compare this `messageSizeThreshold` to the `batchMsgSize` (`batchMsgSize < c.messageSizeThreshold`). This comparison I don't think will handle this scenario as nicely.
> 
> Could you make the tweaks to support this second scenario?



> Thanks @malimohub, glad you are using it and appreciate you contributing back this to the library! I made a couple small syntax suggestions for now, but after reading through I think there is still something missing with the update to `SendMessageBatch`.
> 
> There's a couple scenarios here that we'd like to cover:
> 
> 1. If the entire **batch** size exceeds 256 KB, then send **ALL** records to S3 (in the future this might be tweaked to only send the messages that cause it to exceed this limit). You appear to cover this scenario already!
> 2. If an **individual** message exceeds the `messageSizeThreshold` (which can be configured), then that **individual** message should be sent to S3 as a pointer. This behavior I believe is slightly muddied because you compare this `messageSizeThreshold` to the `batchMsgSize` (`batchMsgSize < c.messageSizeThreshold`). This comparison I don't think will handle this scenario as nicely.
> 
> Could you make the tweaks to support this second scenario?

@jwcorle thanks for taking a look. Ive added a some code to cover the the case you describe below:
If an individual message exceeds the messageSizeThreshold (which can be configured), then that individual message should be sent to S3 as a pointer. This behavior I believe is slightly muddied because you compare this messageSizeThreshold to the batchMsgSize (batchMsgSize < c.messageSizeThreshold). This comparison I don't think will handle this scenario as nicely

Also added some tests around it.

**Comment by jwcorle** [comment]
@malimohub I think the flow of this method is a little back-and-forth now, instead of going back and forth over comments, I'll plan to push up my thoughts as a commit to this PR later today if that's okay with you?

**Comment by malimohub** [comment]
> @malimohub I think the flow of this method is a little unclear now, while functional, I think we could improve it! Instead of going back and forth over comments, I'll plan to push up my thoughts as a commit to this PR later today if that's okay with you?

sounds good!!

**Comment by jwcorle** [comment]
@malimohub 

I pushed up a couple tweaks here! The largest change is that **only** the messages that cause the batch to exceed the configured max batch size will be converted to S3 pointers. So continuing the example from your PR description- if you send 10 messages, each 30 KB, only the last 2 messages will be sent as S3 pointers (first 8 are < 256 KB).

I think this should lead to a better user experience and makes that method's logic a little easier to follow! Happy to hear your thoughts as well.

File: client_test.go [hunk 1]
```diff
@@ -449,6 +450,103 @@ func TestSendMessageBatch(t *testing.T) {
...
+		},
+		QueueUrl: aws.String("testing_url|override_bucket"),
+	})
+
+	assert.Zero(t, ms3c.Calls)
+	assert.Len(t, msqsc.Calls, 1)
+	assert.Nil(t, err)
+}
+
+func TestSendMessageBatchSizeAboveThreshold(t *testing.T) {
```
Comment (jwcorle) [inline client_test.go:503]: Tweaked this test case- first message is less than the BatchMessageSizeThreshold (`20`), but the second and third are larger than the configured threshold, so are sent as pointers.

**Comment by malimohub** [comment]
@jwcorle this is looking good, i like the smarter batch handling logic!!!

**Comment by jwcorle** [comment]
@malimohub Awesome, I'm happy to merge this in if you think it's good to go.

**Comment by malimohub** [comment]
> @malimohub Awesome, I'm happy to merge this in if you think it's good to go.

Yea lets do it

**Review by jwcorle — APPROVED** [review]

**Pull request merged**